common/src/main/java/io/pravega/common/util/btree/sets/BTreeSetPage.java

```diff
@@ -0,0 +1,879 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util.btree.sets;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.pravega.common.util.ArrayView;
+import io.pravega.common.util.BitConverter;
+import io.pravega.common.util.ByteArrayComparator;
+import io.pravega.common.util.ByteArraySegment;
+import io.pravega.common.util.IllegalDataFormatException;
+import io.pravega.common.util.btree.SearchResult;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+
+/**
+ * Represents a Page (Node) within a {@link BTreeSet}. Pages can be of type {@link IndexPage} or {@link LeafPage}.
+ */
+@NotThreadSafe
+abstract class BTreeSetPage {
+    //region Serialization format
+
+    /**
+     * Format Version related fields. The version itself is the first byte of the serialization. When we will have to
+     * support multiple versions, we will need to read this byte and choose the appropriate deserialization approach.
+     * We cannot use VersionedSerializer in here - doing so would prevent us from efficiently querying and modifying the
```

Comment:
It is not great that we have to create an exception for the serialization versioning. I understand that the concern is the overhead of serializing, but could you elaborate on why you believe this is concern that justifies creating an exception for not using `VersionedSerializer`? Is there a way of using `VersionedSerializer` that could solve the issue, e.g., changing a version such that it is not backwards compatible when the overhead is high?

Reply:
The problem with VersionedSerializer is that it forces me to read the whole input (buffer) and store it in another buffer. That is, it reads the data before it gives it to me. 

In this case, since I'm always dealing with huge chunks of data, reading it (and copying it, since that's what VersionedSerializer does) will add extra CPU and GC overhead that will slow down the whole process.

This reason alone was enough for me to choose not to use the convenience of `VersionedSerializer` and use a version flag instead.

Reply:
I'm concerned that at this point I don't understand when it is acceptable to not comply and version directly instead of using the versioned serializer. One of the main goals of versioned serializer was to make sure that we are versioning persistent and external data structures across Pravega in a consistent manner. 

It might be acceptable to do it due to the performance concern, but I'm wondering what the criteria is for exceptions for other PRs.

Reply:
(copy-pasting offline conversation conclusion)
The way to perform an upgrade with this format is as follows:
- We can create new serialization versions by adding new fields at the end of the current serialization.
- There isn't anything to do at this time in terms of accepting a larger input for BTreeSetPage. The format encodes the length of the BTreeSetPage and the parsing logic validates that the given buffer isn't smaller.
    - The page format is: `{Header}{ItemOffsets}{ItemData}{PageId}`. 
    - The length of an item is calculated by getting the difference between that item's offset and the next item's offset.
    - Then line 152 slices the contents buffer to the size encoded in the header.
- So in case of a backwards compatible upgrade (from Version 1), it would work like this:
    - New code defines Version 2, but does not use it yet (it just understands it)
    - New code will serialize using Version 1, but encode its additional data after the end of the buffer. It will make sure that the size encoded in the header ends where the old code expects it to.
    - Old code will read this size and slice the input, essentially dropping new data.
    - New code will be programmed to read past this (if anything) and process any additional fields.
    - Eventually, in a subsequent release, we will change the code to serialize to Version 2, thus completing the upgrade.

Not as smooth as VersionedSerializer (which was doing all this legwork for us), but this is the tradeoff I'm willing to accept for better perf.